### PR TITLE
[develop < T0026-GA] Wrap query modules

### PR DESCRIPTION
**Pull request by brunos252**
### Description

Wrapper class for query modules.

### Pull request type

- [x] Feature

### Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

######################################

### Reviewer checklist (the reviewer checks this part)
- [ ] Core feature implementation
- [ ] Tests
- [ ] Code documentation
- [ ] Documentation on [memgraph/docs](https://github.com/memgraph/docs)

######################################

### CODE REVIEW

**Comment by brunos252** [comment]
The idea currently is that we add an argument self.query_modules to the memgraph object, we create it upon first calling get_procedures, and fetch it for subsequent calls. Init of self.query_modules could also be done in memgraph.__init__(),
what do you think?

File: gqlalchemy/memgraph.py [hunk 1]
```diff
@@ -543,3 +556,146 @@ def create_relationship(self, relationship: Relationship) -> Optional[Relationsh
         )
 
         return self.get_variable_assume_one(results, "relationship")
+
+    def get_procedures(self, starts_with: str = None, update: bool = False) -> List["QueryModule"]:
```
Comment (katarinasupe) [inline gqlalchemy/memgraph.py:560]: I now remember that you asked about `Optional` the other day. I read a bit about it and I learned that in Python>=3.9 it's the same whether you use `starts_with: str = None` or `starts_with: Optional[str]`. Do you agree or did I get it wrong?

File: gqlalchemy/memgraph.py [hunk 2]
```diff
@@ -543,3 +556,146 @@ def create_relationship(self, relationship: Relationship) -> Optional[Relationsh
...
+        and a list of returns:
+        [{"name": "ret", "type": "STRING"}]
+
+        Dictionary consists of fields: "name" - argument name, "type" - data
+        type of argument and "default" where default argument value is given
+
+        Args:
+            signature: module signature as returned by Cypher CALL operation
+        """
+        end_arguments_parantheses = signature.index(")")
```
Comment (katarinasupe) [inline gqlalchemy/memgraph.py:602]: Correct the spelling in variable name: parentheses*

File: gqlalchemy/memgraph.py [hunk 3]
```diff
@@ -543,3 +556,146 @@ def create_relationship(self, relationship: Relationship) -> Optional[Relationsh
...
+    def set_inputs(self, **kwargs) -> None:
+        """Set values for QueryModule arguments so the module can be called.
+
+        Kwargs:
+            Named arguments in self.arguments.
+
+        Raises:
+            KeyError: Passed an argument not in the self.arguments list.
+        """
+        for kwarg in kwargs.keys():
```
Comment (katarinasupe) [inline gqlalchemy/memgraph.py:666]: I would consider changing variable names, here `kwarg` is not descriptive enough. Maybe `argument_key`.

File: gqlalchemy/memgraph.py [hunk 4]
```diff
@@ -543,3 +556,146 @@ def create_relationship(self, relationship: Relationship) -> Optional[Relationsh
...
+    def get_inputs(self) -> str:
+        """return inputs in form "value1, value2, ..." for QueryBuilder call()
+        method.
+
+        Raises:
+            KeyError: Cannot get all values of arguments because one or more is
+            not set.
+        """
+        arguments_str = ""
+        for arg in self.arguments:
```
Comment (katarinasupe) [inline gqlalchemy/memgraph.py:684]: Maybe again `argument_dict` would be more descriptive.

File: gqlalchemy/memgraph.py [hunk 5]
```diff
@@ -543,3 +556,146 @@ def create_relationship(self, relationship: Relationship) -> Optional[Relationsh
...
+            if QM_KEY_VALUE in arg:
+                val = arg[QM_KEY_VALUE]
+            elif QM_KEY_DEFAULT in arg:
+                val = arg[QM_KEY_DEFAULT]
+            else:
+                raise KeyError(f"{arg[QM_KEY_NAME]} has no value set.")
```
Comment (katarinasupe) [inline gqlalchemy/memgraph.py:690]: Do we have to wrap this error in GQLAchemy errors? Not sure what's the practice, so just asking so that I know too 😄

**Comment by g-despot** [comment]
Could you include a test that uses the `QueryModule` class with the query builder and the `call()` method?

File: gqlalchemy/memgraph.py (hunk 5 above)
Comment (brunos252) [reply gqlalchemy/memgraph.py:690]: From talking first with Toni and then Ivan and Mrma, KeyError is a standard error and doesn't need to be wrapped

**Comment by brunos252** [comment]
Decided to move the functions for parsing out of Memgraph class and to utilities.py module

File: gqlalchemy/memgraph.py [hunk 6]
```diff
@@ -543,3 +552,90 @@ def create_relationship(self, relationship: Relationship) -> Optional[Relationsh
...
+        self.is_write = module_dict[QM_FIELD_IS_WRITE]
+        self.path = module_dict[QM_FIELD_PATH]
+        self.signature = module_dict[QM_FIELD_SIGNATURE]
+        self.arguments = arguments
+        self.returns = returns
+
+    def __str__(self) -> str:
+        return self.name
+
+    def set_inputs(self, **kwargs) -> None:
```
[the remark this answers is not in the document]
Comment (brunos252) [reply gqlalchemy/memgraph.py:597]: yeah, like this it sounds too much like getters and setters to me, I changed it to:
set_argument_values
get_arguments_for_call
What do you think? Are they too long?

File: gqlalchemy/graph_algorithms/query_modules.py [hunk 7]
```diff
@@ -0,0 +1,153 @@
...
+QM_FIELD_IS_EDITABLE = "is_editable"
+QM_FIELD_IS_WRITE = "is_write"
+QM_FIELD_PATH = "path"
+QM_FIELD_SIGNATURE = "signature"
+
+
+class QueryModule:
+    """Class representing a single MAGE query module."""
+
+    def __init__(self, module_dict: Dict) -> None:
```
Comment (BorisTasevski) [inline gqlalchemy/graph_algorithms/query_modules.py:32]: Can we replace this dict with kwargs?

File: gqlalchemy/graph_algorithms/query_modules.py [hunk 8]
```diff
@@ -0,0 +1,153 @@
...
+    def set_argument_values(self, **kwargs) -> None:
+        """Set values for QueryModule arguments so the module can be called.
+
+        Kwargs:
+            Named arguments in self.arguments.
+
+        Raises:
+            KeyError: Passed an argument not in the self.arguments list.
+        """
+        for argument_name in kwargs.keys():
```
Comment (BorisTasevski) [inline gqlalchemy/graph_algorithms/query_modules.py:55]: You can remove `.keys()` the behavior will stay the same

File: gqlalchemy/graph_algorithms/query_modules.py [hunk 9]
```diff
@@ -0,0 +1,153 @@
...
+            has_arg = False
+            for argument_dict in self.arguments:
+                if argument_dict[QM_KEY_NAME] == argument_name:
```
Comment (BorisTasevski) [inline gqlalchemy/graph_algorithms/query_modules.py:58]: use `in` operator here. This for loop will run more times than necessary(what if argument name is found in the first place?)

File: gqlalchemy/graph_algorithms/query_modules.py [hunk 10]
```diff
@@ -0,0 +1,153 @@
...
+    def get_arguments_for_call(self) -> str:
+        """return inputs in form "value1, value2, ..." for QueryBuilder call()
+        method.
+
+        Raises:
+            KeyError: Cannot get all values of arguments because one or more is
+            not set.
+        """
+        arguments_str = ""
+        for argument_dict in self.arguments:
```
Comment (BorisTasevski) [inline gqlalchemy/graph_algorithms/query_modules.py:73]: add newline before loop

File: gqlalchemy/graph_algorithms/query_modules.py [hunk 11]
```diff
@@ -0,0 +1,153 @@
...
+            if argument_dict[QM_KEY_TYPE] == "STRING":
+                arguments_str += '"' + val + '"'
+            else:
+                arguments_str += val
+
+            arguments_str += ", "
+
+        arguments_str = arguments_str[:-2]
+
+        return arguments_str
```
Comment (BorisTasevski) [inline gqlalchemy/graph_algorithms/query_modules.py:90]: `return arguments_str[:-2]`

File: gqlalchemy/graph_algorithms/query_modules.py [hunk 12]
```diff
@@ -0,0 +1,153 @@
...
+    and a list of returns:
+    [{"name": "ret", "type": "STRING"}]
+
+    Dictionary consists of fields: "name" - argument name, "type" - data
+    type of argument and "default" where default argument value is given
+
+    Args:
+        signature: module signature as returned by Cypher CALL operation
+    """
+    end_arguments_parentheses = signature.index(")")
```
Comment (BorisTasevski) [inline gqlalchemy/graph_algorithms/query_modules.py:111]: Can we replace all of the hard coded strings with constants?

File: gqlalchemy/graph_algorithms/query_modules.py [hunk 13]
```diff
@@ -0,0 +1,153 @@
...
+        vars_field: signature field inside parentheses
+    """
+    vars = []
+
+    if len(vars_field) == 0:
+        return vars
+
+    list_of_vars = vars_field.split(", ")
+
+    for var in list_of_vars:
```
Comment (BorisTasevski) [inline gqlalchemy/graph_algorithms/query_modules.py:140]: This for loop can be rewritten. We can discuss it personally

File: gqlalchemy/memgraph.py [hunk 14]
```diff
@@ -543,3 +544,26 @@ def create_relationship(self, relationship: Relationship) -> Optional[Relationsh
...
+
+        Args:
+            starts_with: Return those modules that start with this string.
+            (Optional)
+            update: Whether to update the list of modules in
+            self.query_modules. (Optional)
+        """
+        if not hasattr(self, "query_modules") or update:
+            results = self.execute_and_fetch("CALL mg.procedures() YIELD *;")
+            self.query_modules = []
```
Comment (BorisTasevski) [inline gqlalchemy/memgraph.py:562]: rewrite this as list comprehension

File: gqlalchemy/memgraph.py [hunk 15]
```diff
@@ -543,3 +544,26 @@ def create_relationship(self, relationship: Relationship) -> Optional[Relationsh
...
+            for module_dict in results:
+                self.query_modules.append(QueryModule(module_dict))
+
+        if starts_with is not None:
```
Comment (BorisTasevski) [inline gqlalchemy/memgraph.py:566]: `return self.query_modules if starts_with is None else [q for q in self.query_modules if q.name.startswith(starts_with)]`

**Review by BorisTasevski — CHANGES_REQUESTED** [review]

File: gqlalchemy/graph_algorithms/query_modules.py (hunk 11 above)
Comment (Josipmrden) [reply gqlalchemy/graph_algorithms/query_modules.py:90]: Wouldn't agree. Returning plain result that is already processed is better for debugging

File: gqlalchemy/memgraph.py (hunk 14 above)
Comment (Josipmrden) [reply gqlalchemy/memgraph.py:562]: self.query_modules = [QueryModule(x) for x in results] I guess

File: gqlalchemy/memgraph.py (hunk 15 above)
Comment (Josipmrden) [reply gqlalchemy/memgraph.py:566]: I just don't like returning things without being processed first and having them clear in a variable :joy:

**Review by Josipmrden — CHANGES_REQUESTED** [review]

File: gqlalchemy/memgraph.py (hunk 15 above)
Comment (BorisTasevski) [reply gqlalchemy/memgraph.py:566]: Unfortunately, this is the pythonic way...

File: gqlalchemy/graph_algorithms/query_modules.py (hunk 11 above)
Comment (BorisTasevski) [reply gqlalchemy/graph_algorithms/query_modules.py:90]: Additional assignment is not necessary here, but ok

**Review by Josipmrden — APPROVED** [review]

**Review by katarinasupe — APPROVED** [review]

**Review by BorisTasevski — APPROVED** [review]

**Pull request merged**